renderer/FrameBuffer.cpp

```diff
@@ -464,3 +464,221 @@ void FB_TogglePrimary( bool on ) {
+	frameBuffers.currentDraw = nullptr;
+	frameBuffers.currentRead = nullptr;
+}
+
+FrameBuffer::FrameBuffer( GLuint fbo, GLuint width, GLuint height, int msaa )
+	: fbo( fbo ), colorBuffer( 0 ), depthStencilBuffer( 0 ), width( width ), height( height ), msaa( msaa ) {
+}
+
+
+FrameBuffer::~FrameBuffer() {
+	if( fbo != 0 ) {
+		qglDeleteFramebuffers( 1, &fbo );
+	}
+	if( colorBuffer != 0 ) {
+		qglDeleteRenderbuffers( 1, &colorBuffer );
+	}
+	if( depthStencilBuffer != 0 ) {
+		qglDeleteRenderbuffers( 1, &depthStencilBuffer );
+	}
+}
+
+FrameBuffer * FrameBuffer::Create( GLuint width, GLuint height, int msaa ) {
+	GLuint fbo = 0;
+	qglGenFramebuffers( 1, &fbo );
+	if( fbo == 0 ) {
+		common->FatalError( "Failed to allocate framebuffer object" );
+	}
+
+	FrameBuffer *frameBuffer = new FrameBuffer( fbo, width, height, msaa );
+	frameBuffers.allFrameBuffers.push_back( frameBuffer );
+	common->Printf( "FBO created: %dx%d, AA = %d\n", width, height, msaa );
+	return frameBuffer;
+}
+
+FrameBuffer * FrameBuffer::BackBuffer() {
+	if( frameBuffers.backBuffer == 0 ) {
+		frameBuffers.backBuffer = new FrameBuffer( 0, glConfig.vidWidth, glConfig.vidHeight, 0 );
+		frameBuffers.allFrameBuffers.push_back( frameBuffers.backBuffer );
+		common->Printf( "Backbuffer dimensions: %dx%d\n", frameBuffers.backBuffer->width, frameBuffers.backBuffer->height );
+	}
+	return frameBuffers.backBuffer;
+}
+
+void FrameBuffer::CreateColorBuffer() {
+	if( colorBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &colorBuffer );
+		if( colorBuffer == 0 ) {
+			common->FatalError( "Failed to allocate color render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, colorBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_RGBA, width, height );
+	} else {
+		qglRenderbufferStorage( GL_RENDERBUFFER, GL_RGBA, width, height );
+	}
+
+	Bind();
+	qglFramebufferRenderbuffer( GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, colorBuffer );
+}
+
+void FrameBuffer::CreateDepthStencilBuffer() {
+	if( depthStencilBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &depthStencilBuffer );
+		if( depthStencilBuffer == 0 ) {
+			common->FatalError( "Failed to allocate depth/stencil render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, depthStencilBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_DEPTH24_STENCIL8, width, height );
+	} else {
+		qglRenderbufferStorage( GL_RENDERBUFFER, GL_DEPTH24_STENCIL8, width, height );
+	}
+
+	Bind();
+	qglFramebufferRenderbuffer( GL_FRAMEBUFFER, GL_DEPTH_STENCIL_ATTACHMENT, GL_RENDERBUFFER, depthStencilBuffer );
+}
+
+void FrameBuffer::AddColorTexture( idImage *colorTexture ) {
+	if( msaa > 1 ) {
+		common->FatalError( "Rendering to texture for a framebuffer with AA enabled is not currently supported." );
+	}
+	if( colorBuffer != 0 ) {
+		common->Warning( "Setting framebuffer color attachment to texture, but a color render buffer was previously created." );
```

Comment:
Doesn't it make sense to put assert here?
I think this case is no less serious than the one just above which throws FatalError. It shows a clear logical error somewhere in renderer, and it is quite unlikely that things will work well afterwards. In comparison, ignoring `msaa` can have no cnosequences except for lack of antialiasing.

Reply:
Well, it's a bit of a fuzzy line. Technically, reassigning the color attachment to a texture after the fact will still leave the FBO functional, so not really a fatal error, but it's probably wrong. So I went with a warning here. Possibly asserts would be the better alternative.

The MSAA case is a bit trickier. If you mix an AA renderbuffer with a non-AA texture, you will end up with an invalid framebuffer, so that's closer to a real error. Of course, if you assign textures everywhere, it would end up working. But yeah, probably should make the error handling consistent.